File: python/62. Unique Paths.py
class Solution(object):
    def uniquePaths(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        if n < 0 or m < 0:
            return 0
        total = m + n - 2
        upper = lower = 1
        for i in range(m - 1):
            upper *= total
            total -= 1
        for i in range(1, m):
            lower *= i
        return upper/lower

File: cpp/51-100/62. Unique Paths.cpp
class Solution {
public:
	int uniquePaths(int m, int n) {
		if(m > n)
			return uniquePaths(n, m);
		vector<int> pre(m + 1, 1);
		vector<int> cur(m + 1, 0);
		pre[0] = 0;
		for(int i = 0; i < n; i++)
		{
			for(int j = 1; j <= m; j++)
			{
				cur[j] = cur[j-1] + pre[j];
			}
			swap(cur, pre);
		}
		return cur[m];
	}
};